test/integration/cdi/example-driver/app/kubeletplugin.go

```diff
@@ -0,0 +1,161 @@
+*/
+package app
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	cdiapi "github.com/container-orchestrated-devices/container-device-interface/pkg/cdi"
+	specs "github.com/container-orchestrated-devices/container-device-interface/specs-go"
+	"k8s.io/klog/v2"
+	drapbv1 "k8s.io/kubernetes/pkg/kubelet/apis/dra/v1alpha1"
+	"k8s.io/kubernetes/test/integration/cdi/example-driver/kubeletplugin"
+	plugin "k8s.io/kubernetes/test/integration/cdi/example-driver/kubeletplugin"
+)
+
+const (
+	deviceName = "exampledevice"
+	cdiVersion = "0.2.0"
+)
+
+type examplePlugin struct {
+	logger klog.Logger
+	d      plugin.DRAPlugin
+
+	cdiDir     string
+	driverName string
+}
+
+var _ drapbv1.NodeServer = &examplePlugin{}
+
+// runPlugin sets up the plugin and waits for connections from kubelet.
+func runPlugin(logger klog.Logger, cdiDir, driverName, endpoint, draAddress, pluginRegistrationPath string) error {
+	plugin := examplePlugin{
+		logger:     logger,
+		cdiDir:     cdiDir,
+		driverName: driverName,
+	}
+	if err := plugin.start(driverName, endpoint, draAddress, pluginRegistrationPath); err != nil {
+		return fmt.Errorf("start example plugin: %v", err)
+	}
+	// TODO: graceful termination via signal. Needs to be implemented in the caller.
+	select {}
+	plugin.stop()
+	return nil
+}
+
+// getJSONFilePath returns the absolute path where CDI file is/should be.
+func (ex *examplePlugin) getJSONFilePath(claimUID string) string {
+	return filepath.Join(ex.cdiDir, fmt.Sprintf("%s-%s.json", ex.driverName, claimUID))
+}
+
+// start sets up the servers that are necessary for a DRA kubelet plugin.
+func (ex *examplePlugin) start(driverName, endpoint, draAddress, pluginRegistrationPath string) error {
+	// Ensure that directories exist, creating them if necessary. We want
+	// to know early if there is a setup problem that would prevent
+	// creating those directories.
+	if err := os.MkdirAll(ex.cdiDir, os.FileMode(0750)); err != nil {
+		return fmt.Errorf("create CDI directory: %v", err)
+	}
+	if err := os.MkdirAll(filepath.Dir(endpoint), 0750); err != nil {
+		return fmt.Errorf("create socket directory: %v", err)
+	}
+
+	d, err := kubeletplugin.Start(ex.logger, driverName, endpoint, draAddress, pluginRegistrationPath, ex)
+	if err != nil {
+		return fmt.Errorf("start kubelet plugin: %v", err)
+	}
+	ex.d = d
+
+	return nil
+}
+
+// stop ensures that all servers are stopped and resources freed.
+func (ex *examplePlugin) stop() {
+	ex.d.Stop()
+}
+
+// NodePrepareResource ensures that the CDI file for the claim exists. It uses
+// a deterministic name to simplify NodeUnprepareResource (no need to remember
+// or discover the name) and idempotency (when called again, the file simply
+// gets written again).
+func (ex *examplePlugin) NodePrepareResource(ctx context.Context, req *drapbv1.NodePrepareResourceRequest) (*drapbv1.NodePrepareResourceResponse, error) {
+	logger := klog.FromContext(ctx)
+	kind := ex.driverName + "/" + deviceName
```

Comment:
This is still a questionable solution from my point of view. Considering that driver name can be set from command line and we're not validating it, we should avoid using in the kind in my opinion.

Btw, @klihub calls second part of the kind a device class, e.g. intel.com/fpga consists of a vendor (intel.com) and device class (fpga).

Reply:
https://github.com/container-orchestrated-devices/container-device-interface/blob/main/SPEC.md#kind:

"kind": "field specifies a label which uniquely identifies the device vendor ... The kind label has two segments: a prefix and a name, separated by a slash (/)."

"The prefix must be a DNS subdomain": we don't validate it, but the driver name must satisfy this, so this is okay. If user choose something else, they get garbage in/garbage out.

"The name segment is required and must be 63 characters or less, beginning and ending with an alphanumeric character ..." - UID would be fine here.

I'll switch to this:
```
	deviceName := "claim-" + req.ClaimUid
	kind := ex.driverName + "/" + deviceName
```

Reply:
> we don't validate it, but the driver name must satisfy this

why? if it must satisfy it we should validate it I guess.

Reply:
In quality code, yes. In a test or demo driver not so much.

Reply:
I just don't understand the benefit of this. Why we can't use k8s.io as a vendor? Why we should use something passed  to us as a command line parameter?

Reply:
I'd propose to use constant kind k8s.io/test or something like that, where test is a class of devices served by the plugin.

Reply:
> I just don't understand the benefit of this. Why we can't use k8s.io as a vendor? Why we should use something passed to us as a command line parameter?

Debugging will be easier if different driver instances on the same cluster use different kind strings. We will have that situation once the test driver is used for E2E testing and each test instantiates its own driver (necessary to avoid inter-dependencies between tests).

> I'd propose to use constant kind k8s.io/test or something like that, where test is a class of devices served by the plugin.

Doesn't the name have to be unique? Hmm, no. It's the combination of kind+device name which has to be unique.

Reply:
I've switched to "kind=<driver name>/claim" and "device name=claim-<uid>".

Reply:
Another parameter is redundant. But as you seem to be so concerned about the lack of validation, let's add that for the driver name. None of the CSI sidecars have that although it's also required, but we can of course always do better.

My only concern is that "do better" still comes with a cost, even if it is small in this case. Done.

Reply:
I was mostly concerned about using driver name as a kind. It's not the same. But I agree, for the test driver it's not so important.